### PR TITLE
[GlobalISel] Import extract/insert subvector

### PR DESCRIPTION
**Pull request by tschuett**
Test: AArch64/GlobalISel/irtranslator-subvector.ll

Reference:

https://llvm.org/docs/LangRef.html#llvm-vector-extract-intrinsic 
https://llvm.org/docs/LangRef.html#llvm-vector-insert-intrinsic

### CODE REVIEW

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 1]
```diff
@@ -2588,6 +2588,20 @@ bool IRTranslator::translateKnownIntrinsic(const CallInst &CI, Intrinsic::ID ID,
...
                          getOrCreateVReg(*CI.getOperand(1)));
     return true;
+  case Intrinsic::vector_extract: {
+    ConstantInt *Index = cast<ConstantInt>(CI.getOperand(1));
+    MIRBuilder.buildExtractSubvector(getOrCreateVReg(CI),
+                                     getOrCreateVReg(*CI.getOperand(0)),
+                                     Index->getZExtValue());
+    return true;
+  }
+  case Intrinsic::vector_insert: {
```
Comment (qcolombet) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:2598]: What would it take to reuse the implementation of `IRTranslator::translateInsertElement` directly? (And same question for extract)

Reimplementing things here will lead us to having to fix the same issues we had in the past.
E.g.,
```
  // If it is a <1 x Ty> vector, use the scalar as it is
  // not a legal vector type in LLT.
  if (auto *FVT = dyn_cast<FixedVectorType>(U.getType());
      FVT && FVT->getNumElements() == 1)
    return translateCopy(U, *U.getOperand(1), MIRBuilder);
```

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:2598]: `G_INSERT_SUBVECTOR` takes two vectors and an index and returns a vector. The second argument is inserted into the first argument. If the second vector would be a `<1 x Ty> vector`, we would still need to do a `G_INSERT_VECTOR_ELT`.

Comment (qcolombet) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:2598]: Ah right, missed it was subvector.

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 2]
```diff
@@ -3163,6 +3167,58 @@ bool IRTranslator::translateInsertElement(const User &U,
...
+    if (auto *InputType = dyn_cast<FixedVectorType>(U.getOperand(0)->getType());
+        InputType && InputType->getNumElements() == 1) {
+      // We are inserting an illegal fixed vector into an illegal
+      // fixed vector, use the scalar as it is not a legal vector type
+      // in LLT.
+      return translateCopy(U, *U.getOperand(0), MIRBuilder);
+    }
+    if (auto *InputType =
+            dyn_cast<FixedVectorType>(U.getOperand(0)->getType())) {
+      // We are inserting an illegal fixed vector into a fixed vector, use the
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:3197]: InputType not used, is this missing the InputType && InputType->getNumElements() == 1 like above?

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 3]
```diff
@@ -3163,6 +3167,58 @@ bool IRTranslator::translateInsertElement(const User &U,
...
+      // scalar as it is not a legal vector type in LLT.
+      Register Idx = getOrCreateVReg(*CI);
+      MIRBuilder.buildInsertVectorElement(Dst, Vec, Elt, Idx);
+      return true;
+    }
+    if (auto *InputType =
+            dyn_cast<ScalableVectorType>(U.getOperand(0)->getType())) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:3204]: Same here

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 4]
```diff
@@ -3191,6 +3247,55 @@ bool IRTranslator::translateExtractElement(const User &U,
...
+    if (auto *InputType =
+            dyn_cast<FixedVectorType>(U.getOperand(0)->getType())) {
+      // We are extracting a fixed vector from a fixed vector, use the
+      // scalar as it is not a legal vector type in LLT.
+      Register Idx = getOrCreateVReg(*CI);
+      MIRBuilder.buildExtractVectorElement(Res, Vec, Idx);
+      return true;
+    }
+    if (auto *InputType =
+            dyn_cast<ScalableVectorType>(U.getOperand(0)->getType())) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:3281]: Same here

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp [hunk 5]
```diff
@@ -3191,6 +3247,55 @@ bool IRTranslator::translateExtractElement(const User &U,
...
+  if (auto *ResultType = dyn_cast<FixedVectorType>(U.getType());
+      ResultType && ResultType->getNumElements() == 1) {
+    if (auto *InputType = dyn_cast<FixedVectorType>(U.getOperand(0)->getType());
+        InputType && InputType->getNumElements() == 1) {
+      // We are extracting an illegal fixed vector from an illegal fixed vector,
+      // use the scalar as it is not a legal vector type in LLT.
+      return translateCopy(U, *U.getOperand(0), MIRBuilder);
+    }
+    if (auto *InputType =
+            dyn_cast<FixedVectorType>(U.getOperand(0)->getType())) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:3273]: Same here

File: llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp (hunk 2 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp:3196]: Nope. It is meant to say `InputType && InputType->getNumElements() != 1`.

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-subvector.ll [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+  ; CHECK-NEXT:   [[C:%[0-9]+]]:_(s32) = G_CONSTANT i32 1
+  ; CHECK-NEXT:   [[COPY4:%[0-9]+]]:_(s32) = COPY [[UV]](s32)
+  ; CHECK-NEXT:   G_STORE [[COPY4]](s32), [[COPY3]](p0) :: (store (s32) into %ir.p, align 16)
+  ; CHECK-NEXT:   $w0 = COPY [[C]](s32)
+  ; CHECK-NEXT:   RET_ReallyLR implicit $w0
+entry:
+  %vector = call <1  x i32> @llvm.vector.insert.v1i32.v4i32(<1 x i32> %a, <1 x i32> %b, i64 0)
+  store <1 x i32> %vector, ptr %p, align 16
+  ret i32 1
+}
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-subvector.ll:206]: Can you add some tests with FP types, and vectors of pointers?

**Review by arsenm — COMMENTED** [review]
LGTM, but could use some test coverage for different element classes

File: llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-subvector.ll [hunk 7]
```diff
@@ -20,6 +20,47 @@ entry:
...
+  ; CHECK-NEXT:   [[CONCAT_VECTORS:%[0-9]+]]:_(<4 x s64>) = G_CONCAT_VECTORS [[COPY]](<2 x s64>), [[COPY1]](<2 x s64>)
+  ; CHECK-NEXT:   [[COPY2:%[0-9]+]]:_(<2 x s64>) = COPY $q2
+  ; CHECK-NEXT:   [[COPY3:%[0-9]+]]:_(s32) = COPY $w0
+  ; CHECK-NEXT:   [[C:%[0-9]+]]:_(s64) = G_CONSTANT i64 1
+  ; CHECK-NEXT:   [[INSERT_SUBVECTOR:%[0-9]+]]:_(<4 x s64>) = G_INSERT_SUBVECTOR [[CONCAT_VECTORS]], [[COPY2]](<2 x s64>), 0
+  ; CHECK-NEXT:   [[EVEC:%[0-9]+]]:_(s64) = G_EXTRACT_VECTOR_ELT [[INSERT_SUBVECTOR]](<4 x s64>), [[C]](s64)
+  ; CHECK-NEXT:   $d0 = COPY [[EVEC]](s64)
+  ; CHECK-NEXT:   RET_ReallyLR implicit $d0
+entry:
+  %vector = call <4 x double> @llvm.vector.insert.v4double.v2double(<4 x double> %a, <2 x double> %b, i64 0)
```
Comment (arsenm) [inline llvm/test/CodeGen/AArch64/GlobalISel/irtranslator-subvector.ll:39]: Use canonical mangling, f64 not double

**Review by arsenm — APPROVED** [review]
lgtm with nit

**Review by arsenm — APPROVED** [review]

**Comment by tschuett** [comment]
Thanks for review.

**Pull request merged**